core/src/main/java/io/questdb/griffin/engine/orderby/LimitedSizeLongTreeChain.java

```diff
@@ -0,0 +1,452 @@
+            }
+
+            setRef(newNode, chainOffset);
+
+            return newNode;
+        }
+    }
+
+    private void refreshMinMaxNode() {
+        long p;
+
+        if (isfirstN) {
+            p = findMaxNode();
+        } else {//lastN/topN
+            p = findMinNode();
+        }
+
+        minMaxNode = p;
+        minMaxRowId = valueChain.getLong(refOf(p));
+    }
+
+    //remove node and put on freelist (if holds only one value in chain)
+    protected void removeAndCache(long node) {
+        if (hasMoreThanOneValue(node)) {
+            removeMostRecentChainValue(node);//don't change minmax   
+        } else {
+            long nodeToRemove = super.remove(node);
+            clearBlock(nodeToRemove);
+            freeList.add(nodeToRemove);//keep node on freelist to minimize allocations
+
+            minMaxRowId = -1;//re-compute after inserting, there's no point doing it now
+            minMaxNode = -1;
+        }
+
+        currentValues--;
+    }
+
+    private void clearBlock(long position) {
+        setParent(position, -1);
+        setLeft(position, -1);
+        setRight(position, -1);
+        setColor(position, BLACK);
+        //assume there's only one value in the chain (otherwise node shouldn't be deleted)
+        long refOffset = refOf(position);
+        assert valueChain.getLong(refOffset + 8) == CHAIN_END;
+        valueChain.putLong(refOffset, FREE_SLOT);
+    }
+
+    private void removeMostRecentChainValue(long node) {
+        long ref = refOf(node);
+        long previousOffset = valueChain.getLong(ref + 8);
+        setRef(node, previousOffset);
+        chainFreeList.add(ref);
+    }
+
+    private boolean hasMoreThanOneValue(long position) {
+        long ref = refOf(position);
+        long previousValue = valueChain.getLong(ref + 8);
+        return previousValue != CHAIN_END;
+    }
+
+    @Override
+    protected void putParent(long value) {
+        root = allocateBlock(-1, value);
+    }
+
+    public class TreeCursor {
+
+        private long treeCurrent;
+        private long chainCurrent;
+
+        public boolean hasNext() {
+            if (chainCurrent != -1) {
+                return true;
+            }
+
+            treeCurrent = successor(treeCurrent);
+            if (treeCurrent == -1) {
+                return false;
+            }
+
+            chainCurrent = refOf(treeCurrent);
+            return true;
+        }
+
+        public long next() {
+            long result = chainCurrent;
+            chainCurrent = valueChain.getLong(chainCurrent + 8);
+            return valueChain.getLong(result);
+        }
+
+        public void toTop() {
+            setup();
+        }
+
+        private void setup() {
+            long p = root;
+            if (p != -1) {
+                while (leftOf(p) != -1) {
+                    p = leftOf(p);
+                }
+            }
+            chainCurrent = refOf(treeCurrent = p);
+        }
+    }
+
```

Comment:
There are several ways to implement this : 
1. ignoring records in cursor (like in LimitRecordCursor) . Fastest approach if number of records is small and/or cursor won't be traversed many times .  
2. removing first/last N records from tree (implemented here) . It doesn't release any memory though . 
3. same as 2 but without fixing tree structure (could be faster if number of records to skip is large because once result is materialized it probably doesn't need to be a balanced tree anymore ) .